pkg/apis/resource/types.go

```diff
@@ -265,45 +247,37 @@ type Device struct {
 	// +required
 	Name string
 
-	// Basic defines one device instance.
-	//
-	// +optional
-	// +oneOf=deviceType
-	Basic *BasicDevice
-}
-
-// BasicDevice defines one device instance.
-type BasicDevice struct {
 	// Attributes defines the set of attributes for this device.
 	// The name of each attribute must be unique in that set.
 	//
-	// The maximum number of attributes and capacities combined is 32.
+	// The maximum number of attributes, capacities and counters combined is 32.
 	//
 	// +optional
 	Attributes map[QualifiedName]DeviceAttribute
 
 	// Capacity defines the set of capacities for this device.
 	// The name of each capacity must be unique in that set.
 	//
-	// The maximum number of attributes and capacities combined is 32.
+	// The maximum number of attributes, capacities and counters combined is 32.
 	//
 	// +optional
 	Capacity map[QualifiedName]DeviceCapacity
 
-	// ConsumesCounter defines a list of references to sharedCounters
+	// ConsumesCounters defines a list of references to sharedCounters
 	// and the set of counters that the device will
 	// consume from those counter sets.
 	//
 	// There can only be a single entry per counterSet.
 	//
-	// The maximum number of device counter consumption entries
-	// is 32. This is the same as the maximum number of shared counters
-	// allowed in a ResourceSlice.
+	// The total number of device counter consumption entries
+	// must be <= 32. In addition, the total number in the
+	// entire ResourceSlice must be <= 1024 (for example,
+	// 64 devices with 16 counters each).
```

Comment:
This resolves https://github.com/kubernetes/kubernetes/pull/129970/files#r2008408292 (adding a link because that thread is soon going to be considered "outdated" by GitHub).